src/coprocessor/dag/expr/compare.rs

```diff
@@ -161,6 +164,115 @@ impl FnCall {
         do_coalesce(self, |v| v.eval_json(ctx, row))
     }
 
+    pub fn greatest_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        do_get_extremum(self, |e| e.eval_int(ctx, row), max)
+    }
+
+    pub fn greatest_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        do_get_extremum(self, |e| e.eval_real(ctx, row), |x, y| x.max(y))
+    }
+
+    pub fn greatest_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        do_get_extremum(self, |e| e.eval_decimal(ctx, row), max)
+    }
+
+    pub fn greatest_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut greatest = Time::new(
+            ctx.cfg.tz.timestamp(mysql::time::ZERO_TIMESTAMP, 0),
+            mysql::types::DATETIME,
+            mysql::DEFAULT_FSP,
+        )?;
+
+        for exp in &self.children {
+            let val = try_opt!(exp.eval_string(ctx, row));
+            let s = str::from_utf8(&val)?;
+            match Time::parse_datetime(s, self.tp.get_decimal() as i8, &ctx.cfg.tz) {
+                Ok(t) => greatest = max(greatest, t),
+                Err(_) => {
+                    if let Err(e) = ctx.handle_invalid_time_error(Error::invalid_time_format(s)) {
+                        return Err(e);
```

Comment:
How about `invalid fsp`?

Reply:
Sorry but I don't get your point. Did you mean an invalid_fsp error?(like https://github.com/pingcap/tikv/blob/c2f348a7dcbc6da64bae960a738dbc6375b26eb6/src/coprocessor/codec/mysql/mod.rs#L32)

Reply:
yes. And it seems `tidb` parse fsp from s while you use `self.tp.get_decimal()` instead?

Reply:
now fsp is parsed from `Time::parse_fsp`(consistent with TiDB). The fsp is valid always so there will no `invalid_fsp` Error, PTAL @AndreMouche